packages/core/streams/langchain-stream.ts

```diff
@@ -3,20 +3,77 @@ import { type AIStreamCallbacks, createCallbacksTransformer } from './ai-stream'
 export function LangChainStream(callbacks?: AIStreamCallbacks) {
   const stream = new TransformStream()
   const writer = stream.writable.getWriter()
+
+  const runs = new Set()
+
+  const handleError = async (e: Error, runId: string) => {
+    runs.delete(runId)
+    await writer.ready
+    await writer.abort(e)
+  }
+
+  const handleStart = async (runId: string) => {
+    runs.add(runId)
+  }
+
+  const handleEnd = async (runId: string) => {
+    runs.delete(runId)
+
+    if (runs.size === 0) {
+      await writer.ready
+      await writer.close()
+    }
+  }
+
   return {
     stream: stream.readable.pipeThrough(createCallbacksTransformer(callbacks)),
     handlers: {
       handleLLMNewToken: async (token: string) => {
         await writer.ready
         await writer.write(token)
       },
-      handleChainEnd: async () => {
-        await writer.ready
-        await writer.close()
+      handleLLMStart: async (_llm: any, _prompts: string[], runId: string) => {
+        handleStart(runId)
       },
-      handleLLMError: async (e: any) => {
-        await writer.ready
-        await writer.abort(e)
+      handleLLMEnd: async (_output: any, runId: string) => {
+        await handleEnd(runId)
+      },
+      handleLLMError: async (e: Error, runId: string) => {
+        await handleError(e, runId)
+      },
+      handleChatModelStart: async (
+        _llm: any,
+        _messages: any,
+        runId: string
+      ) => {
+        handleStart(runId)
+      },
+      handleChatModelEnd: async (_output: any, runId: string) => {
```

Comment:
This callback event does not exist (currently), handleLLMEnd is used by chat models